### PR TITLE
Python3 dev branch pull request

### PR DESCRIPTION
**Pull request by Nandan307**
Acknowledged.

-Nandan

### CODE REVIEW

File: ovs_performance.py [hunk 1]
```diff
@@ -2485,7 +2494,7 @@ def get_ovs_version():
                                 die_on_error=True)
 
     m = re.search('.*([0-9]+.[0-9]+.[0-9]+).*',
-                  result.output)
+                  result.output.decode("utf-8", "ignore"))
     if m:
         return m.group(1)
```
Comment (chaudron) [inline ovs_performance.py:3327]: change this to return str(m.group(1)) or else the  StrictVersion(ovs_version) parts do not work. Please test this without the --no-bridge-config.

File: ovs_performance.py [hunk 2]
```diff
@@ -67,6 +67,7 @@
 import os
 import logging
 import numpy as np
+import pdb
```
Comment (chaudron) [inline ovs_performance.py:70]: I assume this was a leftover and should be removed.

File: ovs_performance.py [hunk 3]
```diff
@@ -816,7 +821,7 @@ def get_active_datapath_flows():
         cmd = 'sh -c "ovs-appctl dpctl/dump-flows system@ovs-system | wc -l"'
 
     result = dut_shell.dut_exec(cmd, die_on_error=True)
-    return int(result.stdout_output)
+    return int(result.stdout_output.decode("utf-8", "ignore"))
```
Comment (chaudron) [inline ovs_performance.py:832]: I think we should fix this in the dut_ssh_shell.py module, to always return utf-8, so all changes in this file can be reverted.

File: ovs_performance.py [hunk 4]
```diff
@@ -1818,9 +1826,9 @@ def get_of_port_packet_stats(of_port, **kwargs):
         sys.exit(-1)
 
     slogger.debug("OF port {0} stats: tx = {1}, tx_drop = {2}, rx = {3}, tx_drop = {3}".
-                  format(of_port, long(tx), long(tx_drop), long(rx), long(rx_drop)))
+                  format(of_port, int(tx), int(tx_drop), int(rx), int(rx_drop)))
```
Comment (chaudron) [inline ovs_performance.py:2657]: Maybe when typecasting twice, we should move the typecast to the assignments above?

File: ovs_performance.py [hunk 5]
```diff
@@ -2499,9 +2508,9 @@ def get_ovs_version():
 def get_ovs_datapath():
     result = dut_shell.dut_exec('sh -c "ovs-appctl dpif/show"',
                                 die_on_error=True)
-    output = result.output.replace("\n", "")
+    output = result.output.replace(b"\n",b"")
```
Comment (chaudron) [inline ovs_performance.py:3361]: Guess a space after ,?

File: ovs_performance.py [hunk 6]
```diff
@@ -3072,12 +3081,13 @@ def main():
     else:
         raise ValueError("No support for this protocol!!")
 
-    with open(csv_file, 'wb') as csvfile:
+    with open(csv_file, 'w') as csvfile:
```
Comment (chaudron) [inline ovs_performance.py:3941]: Any specific reason for changing this? guess window machines now add cr/lf.

File: ovs_performance.py [hunk 7]
```diff
@@ -3072,12 +3081,13 @@ def main():
...
         csv_handle = csv.writer(csvfile, dialect='excel')
 
-        csv_handle.writerow(["Physical port, \"{}\", speed {} Gbit/s".
-                             format(config.physical_interface,
-                                    phy_speed / 1000000000)])
+        csv_handle.writerow(["Physical port"] +
+                            [str(config.physical_interface)] +
```
Comment (chaudron) [inline ovs_performance.py:3945]: Why use two different formats options here? Why not keep it as is?

File: traffic_generator_trex.py [hunk 8]
```diff
@@ -52,8 +52,7 @@
     os.path.dirname(os.path.realpath(__file__)),
     "./trex_stl_lib/external_libs/"))
 
-from trex_stl_lib.api import STLClient, STLError, STLPktBuilder, STLStream, \
-    STLTXCont, STLVmFixIpv4, STLVmFlowVar, STLVmWrFlowVar
+from trex_stl_lib.api import STLClient, STLError, STLPktBuilder, STLStream, STLTXCont, STLVmFixIpv4, STLVmFlowVar, STLVmWrFlowVar
```
Comment (chaudron) [inline traffic_generator_trex.py:55]: Why remove the line break and make this an extra long line?

File: traffic_generator_trex.py [hunk 9]
```diff
@@ -189,10 +188,10 @@ def configure_traffic_stream(self, traffic_flows, nr_of_flows,
                     dst="2.0.0.0")
             L4 = UDP(chksum=0)
 
-            if (len(str(L2/L3/L4)) + 4) > packet_size:  # +4 for Ethernet CRC
-                raise ValueError("Packet size ({} bytes) to small for"
-                                 "requested packet ({} bytes)!".
-                                 format(packet_size, len(L2/L3/L4) + 4))
+            #if (len(str(L2/L3/L4)) + 4) > packet_size:  # +4 for Ethernet CRC
```
Comment (chaudron) [inline traffic_generator_trex.py:191]: Why did this get removed, it should be there...

**Review by chaudron — CHANGES_REQUESTED** [review]

File: ovs_performance.py (hunk 6 above)
Comment (Nandan307) [reply ovs_performance.py:3941]: The unicode error ("wb' was writing bytes to the object whereas 'w' writes as string)

**Comment by chaudron** [comment]
It does not apply cleanly, can you rebase your branch on dev/python3 before submitting?

File: dut_ssh_shell.py [hunk 10]
```diff
@@ -46,7 +46,8 @@ def __init__(self, return_code, stdout_output, stderr_output):
 
     @property
     def output(self):
-        return self.stdout_output + self.stderr_output
+        s = self.stdout_output.decode('utf-8', 'ignore') + self.stderr_output.decode('utf-8', 'ignore')
```
Comment (chaudron) [inline dut_ssh_shell.py]: Here you change the output object only? However, in your previous set, you had changes like "result.stdout_output.decode("utf-8", "ignore")", so I guess you also need to change line 44 and 45?

File: ovs_performance.py [hunk 11]
```diff
@@ -947,8 +962,7 @@ def start_traffic_loop_on_vm(vm, pci):
     cpu_mask = ((1 << (config.dut_vm_nic_queues + 1)) - 1)
     pmd_cpu_mask = cpu_mask & ~0x1
     mac_swap = " --forward-mode=macswap" if config.mac_swap else ""
-    disable_hw_vlan = " --disable-hw-vlan" if vm_dpdk_version < \
-                      StrictVersion('18.2.0') else ""
+    disable_hw_vlan = " --disable-hw-vlan" if vm_dpdk_version < StrictVersion('18.2.0') else ""
```
Comment (chaudron) [inline ovs_performance.py]: Why the change? This moves is over 80 chars?

File: ovs_performance.py [hunk 12]
```diff
@@ -989,12 +1003,11 @@ def get_traffic_rx_stats_from_vm(vm, **kwargs):
     result = dut_shell.dut_exec('', raw_cmd=['sh', '-c', cmd], die_on_error=True)
 
     pkt_rates = [int(re.sub(r'^\s*Rx-pps:\s*', '', s))
-                 for s in re.findall(r'^\s*Rx-pps:\s*\d+$', result.stdout_output,
-                                     re.MULTILINE)]
+                 for s in re.findall(r'^\s*Rx-pps:\s*\d+$', result.stdout_output.decode("utf-8", "ignore"), re.MULTILINE)]
```
Comment (chaudron) [inline ovs_performance.py]: This should be moved out, do not want to see any of these decodes () in ovs_performance.py

File: ovs_performance.py [hunk 13]
```diff
@@ -2443,7 +2472,7 @@ def get_cpu_monitoring_stats():
     #                    Average:   0        -   6982     0.00       0.05       0.00        0.05     -  |__ovs-vswitchd
     regex = re.compile("^Average:\s+[0-9]+\s+-\s+[0-9]+\s+[0-9\.]+\s+[0-9\.]+\s+[0-9\.]+\s+([0-9\.]+).+", re.MULTILINE)
     ovs_cpu_usage = float(0)
-    for match in regex.finditer(results.stdout_output):
+    for match in regex.finditer(results.stdout_output.decode("utf-8","ignore")):
```
Comment (chaudron) [inline ovs_performance.py]: see above

File: ovs_performance.py [hunk 14]
```diff
@@ -2463,7 +2492,8 @@ def get_cpu_monitoring_stats():
     #  %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
     regex = re.compile("^Average:\s+[0-9]+\s+([0-9\.]+)\s+([0-9\.]+)\s+([0-9\.]+)\s+([0-9\.]+)\s+([0-9\.]+)\s+([0-9\.]+)\s+([0-9\.]+)\s+([0-9\.]+)\s+([0-9\.]+)\s+([0-9\.]+)$",
                        re.MULTILINE)
-    for match in regex.finditer(results.stdout_output):
+
+    for match in regex.finditer(results.stdout_output.decode('utf-8')):
```
Comment (chaudron) [inline ovs_performance.py:3283]: see above

File: ovs_performance.py [hunk 15]
```diff
@@ -1347,13 +1360,14 @@ def create_ovs_l3_of_rules(number_of_flows, src_port, dst_port, **kwargs):
 
         if int(result.stdout_output) != total_nr_of_flows:
             lprint("ERROR: Only {0} flows should exsits, but there are {1}!".
-                   format(number_of_flows, int(result.stdout_output)))
-            sys.exit(-1)
+                   format(number_of_flows, int(result.stdout_output, 5)))
+
```
Comment (chaudron) [inline ovs_performance.py]: sys.exit(-1) should not be removed!!

File: ovs_performance.py [hunk 16]
```diff
@@ -1347,13 +1360,14 @@ def create_ovs_l3_of_rules(number_of_flows, src_port, dst_port, **kwargs):
...
 
 #
 # Add OVS Bidirectional L3 OpenFlow rules
 #
 def create_ovs_bidirectional_l3_of_rules(number_of_flows, src_port, dst_port, **kwargs):
+
```
Comment (chaudron) [inline ovs_performance.py]: why? Please remove

File: ovs_performance.py [hunk 17]
```diff
@@ -1434,7 +1448,6 @@ def create_ovs_l3_of_slash_16_rules(number_of_flows,
                    format(number_of_flows, int(result.stdout_output)))
             sys.exit(-1)
 
-
```
Comment (chaudron) [inline ovs_performance.py]: why? Should be two enters after function

File: ovs_performance.py [hunk 18]
```diff
@@ -1502,6 +1515,7 @@ def create_ovs_bidirectional_l4_of_rules(number_of_flows, src_port, dst_port, **
 #
 # Add test bridge setup
 #
+
```
Comment (chaudron) [inline ovs_performance.py:2303]: remove

Comment (chaudron) [reply ovs_performance.py:2303]: If you are an emacs user install flycheck.org and the python modules so it will catch all of these. If you use another editor run flake8 manually.

File: ovs_performance.py [hunk 19]
```diff
@@ -1706,8 +1736,8 @@ def create_ovs_vxlan_bridge():
     # working. So we pause here, asking for restart of the VM.
     #
     if dpdk:
-        print "!!! Finished configuring the OVS bridge, please restart the Virtual Machine !!!"
-        raw_input("Press Enter to continue...")
+        print("!!! Finished configuring the OVS bridge, please restart the Virtual Machine !!!")
+        eval(input("Press Enter to continue..."))
```
Comment (chaudron) [inline ovs_performance.py:2526]: You missed undoing this change?

File: ovs_performance.py [hunk 20]
```diff
@@ -2562,7 +2591,7 @@ def get_of_bridge_mac_address(bridge):
     result = dut_shell.dut_exec(command, die_on_error=True)
 
     m = re.search('\s*LOCAL\({0}\): addr:(.*)'.format(bridge),
-                  result.output)
+                  result.output.decode("utf-8", "ignore"))
```
Comment (chaudron) [inline ovs_performance.py:3382]: Forgot to remove

File: ovs_performance.py [hunk 21]
```diff
@@ -3121,12 +3150,13 @@ def main():
...
         raise ValueError("No support for this protocol!!")
 
-    with open(csv_file, 'wb') as csvfile:
+    with open(csv_file, 'w') as csvfile:
         csv_handle = csv.writer(csvfile, dialect='excel')
 
         csv_handle.writerow(["Physical port, \"{}\", speed {} Gbit/s".
                              format(config.physical_interface,
                                     phy_speed / 1000000000)])
+
```
Comment (chaudron) [inline ovs_performance.py]: not needed

File: traffic_generator_trex.py [hunk 22]
```diff
@@ -53,7 +53,7 @@
     "./trex_stl_lib/external_libs/"))
 
 from trex_stl_lib.api import STLClient, STLError, STLPktBuilder, STLStream, \
-    STLTXCont, STLVmFixIpv4, STLVmFlowVar, STLVmWrFlowVar
+                      STLTXCont, STLVmFixIpv4, STLVmFlowVar, STLVmWrFlowVar
```
Comment (chaudron) [inline traffic_generator_trex.py]: Guess formatting should be like this to keep flake8 happy:

from trex_stl_lib.api import STLClient, STLError, STLPktBuilder, STLStream, \
                             STLTXCont, STLVmFixIpv4, STLVmFlowVar, \
                             STLVmWrFlowVar

File: traffic_generator_trex.py (hunk 9 above)
Comment (chaudron) [reply traffic_generator_trex.py:191]: This is still not fixes, guess you missed the comment

File: traffic_generator_xena.py [hunk 23]
```diff
@@ -159,13 +159,20 @@ def _configure_xena_stream(self, stream_id, traffic_flows, offset,
         else:
             raise ValueError("Unsupported traffic type for Xena tester!!!")
 
-        if (len(str(L2/L3/L4)) + 4) > packet_size:  # +4 for Ethernet CRC
-            raise ValueError("Packet size ({} bytes) to small for requested "
-                             "packet ({} bytes)!".
-                             format(packet_size, len(L2/L3/L4) + 4))
+        # if (len(str(L2/L3/L4)) + 4) > packet_size:  # +4 for Ethernet CRC
```
Comment (chaudron) [inline traffic_generator_xena.py]: Same here, this should not be removed!

File: traffic_generator_xena.py [hunk 24]
```diff
@@ -159,13 +159,20 @@ def _configure_xena_stream(self, stream_id, traffic_flows, offset,
...
 
-        packet_hex = '0x' + str(L2/L3/L4).encode('hex')
+        packet_hex = '0x' + str(L2/L3/L4)
 
+        # The next two statements help
+        # to get rid of the following error:
+        # UnicodeDecodeError: 'ascii' codec can't decode byte 0xd1
+        # in position 25: ordinal not in range(128)
+
+        unic = u''
```
Comment (chaudron) [inline traffic_generator_xena.py]: How does this work? You override the packet_hex in line 167...

Comment (chaudron) [reply traffic_generator_xena.py]: Please test your changes with a Xena also!

**Review by chaudron — CHANGES_REQUESTED** [review]
